fontforge/featurefile.c

```diff
@@ -4732,14 +4736,20 @@ static void fea_ParseSubstitute(struct parseState *tok) {
 	    }
 	} else if ( cnt>=1 && tok->type==tk_by ) {
 	    rpl = fea_ParseMarkedGlyphs(tok,false,false,false);
-	    if ( rpl==NULL ) {
+	    fea_ParseTok(tok);
+	    bool next_is_null = tok->type == tk_NULL;
+	    fea_UnParseTok(tok);
+	    if ( rpl==NULL && !next_is_null ) {
 		LogError(_("No substitution specified on line %d of %s"), tok->line[tok->inc_depth], tok->filename[tok->inc_depth] );
 		++tok->err_count;
-	    } else if ( rpl->has_marks ) {
+	    } else if ( !next_is_null && rpl->has_marks ) {
 		LogError(_("No marked glyphs allowed in replacement on line %d of %s"), tok->line[tok->inc_depth], tok->filename[tok->inc_depth] );
 		++tok->err_count;
 	    } else {
-		if ( cnt==1 && rpl->next==NULL ) {
+		if (next_is_null) {
+		    tok->sofar = fea_process_sub_multiple(tok,glyphs,NULL,tok->sofar);
+		    fea_ParseTok(tok);
```

Comment:
Can you help me out with a quick explanation of why we need to consume an extra token here but not in the other cases?

Reply:
The string "NULL" is handled specially by the FEA parser, it is in its own token class, `tk_NULL`. Therefore, it causes all sorts of issues if it comes up in the next run, and this tk_NULL will never be properly consumed any other way. In the normal cases, where the glyph name in question actually resolves to a glyph, the swallowing of the token is was already done during the function `fea_ParseMarkedGlyphs`.

Reply:
OK --

This pattern makes me a bit nervous given that `fea_UnParseTok()` only sets `backedup` and seems to be there to match up with potential further calls to `fea_ParseTok()`. Before we were calling `fea_process_sub_single()`, `fea_process_sub_multiple()`, etc. with the saved state of the *previous* token. Now we're parsing again, setting `backedup` to true, and then calling those functions.

Is the idea that if parsing a NULL only modifies the `type` and those functions never look at the type then everything works out? That would be OK if the next token *is* a "NULL" but what if it's something else? Do we have an exhaustive list of what it could be? And what if the file has a syntax error that makes the token different -- won't we wind up with weird processing here rather than a syntax error message?

Maybe I just need more guidance about what's going on.

Reply:
We already read in what the replacement is going to be into `rpl`. The only valid thing it can be at this point is either `NULL` (in which case we consume the NULL) or something already read into `rpl`. We still get logical errors if they exist in the input feature file.

Setting `backedup` is enough because it causes the next `fea_ParseTok` to fall through.

Reply:
OK, I've looked through the code in a bit more depth. I guess `tok` is only passed into the `_sub_` functions for error reporting purposes. This isn't great given [feature file whitespace handling](http://adobe-type-tools.github.io/afdko/OpenTypeFeatureFileSpecification.html#2b-white-space), which will lead to incorrect error messages in some cases.

However, assuming that's acceptable I would prefer to centralize the logic this way:

```
	    fea_ParseTok(tok);
	    bool next_is_null = tok->type == tk_NULL;
	    if ( !next_is_null )
	        fea_UnParseTok(tok);
```
preferably with a comment about what's going on.

Reply:
Force pushed, thanks.

Reply:
Oh, sorry, the idea of the added guard is to eliminate the need for `fea_ParseTok(tok);` on the next_is_null branch below. I left that part implicit.

Reply:
As much my fault as yours :-) Really tired.

Force pushed again.